massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,659 @@
+                            .verify_signature(&hash, &s_block_header.signature)
+                            .is_ok()
+                } else {
+                    false
+                }
+            }
+        }
+    }
+
+    fn is_valid(&self) -> bool {
+        let (signature_1, signature_2, hash_1, hash_2, public_key) = match self {
+            Denunciation::Endorsement(de) => (
+                de.signature_1,
+                de.signature_2,
+                de.hash_1,
+                de.hash_2,
+                de.public_key,
+            ),
+            Denunciation::BlockHeader(de) => (
+                de.signature_1,
+                de.signature_2,
+                de.hash_1,
+                de.hash_2,
+                de.public_key,
+            ),
+        };
+
+        hash_1 != hash_2
+            && signature_1 != signature_2
+            && public_key.verify_signature(&hash_1, &signature_1).is_ok()
+            && public_key.verify_signature(&hash_2, &signature_2).is_ok()
+    }
+}
+
+/// Create a new Denunciation from 2 SecureShareEndorsement
+impl TryFrom<(&SecureShareEndorsement, &SecureShareEndorsement)> for Denunciation {
+    type Error = DenunciationError;
+
+    fn try_from(
+        (s_e1, s_e2): (&SecureShareEndorsement, &SecureShareEndorsement),
+    ) -> Result<Self, Self::Error> {
+        // Cannot use the same endorsement twice
+        if s_e1 == s_e2 {
+            return Err(DenunciationError::InvalidInput);
+        }
+
+        // In order to create a Denunciation, there should be the same
+        // slot, index & public key
+        if s_e1.content.slot != s_e2.content.slot
+            || s_e1.content.index != s_e2.content.index
+            || s_e1.content_creator_pub_key != s_e2.content_creator_pub_key
+        {
+            return Err(DenunciationError::InvalidInput);
+        }
+
+        // Check sig of s_e2 but with s_e1.public_key, s_e1.slot, s_e1.index
+        let s_e2_hash = EndorsementDenunciation::compute_hash_for_sig_verif(
+            &s_e1.content_creator_pub_key,
+            &s_e1.content.slot,
+            &s_e1.content.index,
+            &s_e2.content,
+        )?;
+
+        s_e1.content_creator_pub_key
+            .verify_signature(&s_e2_hash, &s_e2.signature)?;
+
+        Ok(Denunciation::Endorsement(EndorsementDenunciation {
+            public_key: s_e1.content_creator_pub_key,
+            slot: s_e1.content.slot,
+            index: s_e1.content.index,
+            signature_1: s_e1.signature,
+            signature_2: s_e2.signature,
+            hash_1: *s_e1.id.get_hash(),
+            hash_2: s_e2_hash,
+        }))
+    }
+}
+
+/// Create a new Denunciation from 2 SecureHeader
+impl TryFrom<(&SecuredHeader, &SecuredHeader)> for Denunciation {
+    type Error = DenunciationError;
+
+    fn try_from((s_bh1, s_bh2): (&SecuredHeader, &SecuredHeader)) -> Result<Self, Self::Error> {
+        // Cannot use the same endorsement twice
```

Comment:
isn't a test missing here?

```suggestion
// Cannot use the same endorsement twice
if s_bh1 == s_bh2 {
    return Err(DenunciationError::InvalidInput);
}
```

Reply:
Ok, you may forget this because
`s_bh1.serialized_data == s_bh2.serialized_data` => `s_bh1 == s_bh2`
Maybe I was confused buy the comment

Reply:
Indeed there is no PartialEq impl for SecuredHeader so I decided to check the serialized_data (which is enough in our case)